genesyscloud/routing_email_domain/resource_genesyscloud_routing_email_domain.go

```diff
@@ -0,0 +1,168 @@
+
+	domains, resp, getErr := proxy.getAllRoutingEmailDomains(ctx)
+	if getErr != nil {
+		return nil, util.BuildAPIDiagnosticError(resourceName, fmt.Sprintf("Failed to get routing email domains error: %s", getErr), resp)
+	}
+
+	if domains == nil || len(*domains) == 0 {
+		return resources, nil
+	}
+
+	for _, domain := range *domains {
+		resources[*domain.Id] = &resourceExporter.ResourceMeta{Name: *domain.Id}
+	}
+	return resources, nil
+}
+
+func createRoutingEmailDomain(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*provider.ProviderMeta).ClientConfig
+	proxy := getRoutingEmailDomainProxy(sdkConfig)
+
+	domainID := d.Get("domain_id").(string)
+	subdomain := d.Get("subdomain").(bool)
+	mxRecordStatus := "VALID"
+	if !subdomain {
+		mxRecordStatus = "NOT_AVAILABLE"
+	}
+
+	sdkDomain := platformclientv2.Inbounddomain{
+		Id:             &domainID,
+		SubDomain:      &subdomain,
+		MxRecordStatus: &mxRecordStatus,
+	}
+
+	log.Printf("Creating routing email domain %s", domainID)
+	domain, resp, err := proxy.createRoutingEmailDomain(ctx, &sdkDomain)
+	if err != nil {
+		return util.BuildAPIDiagnosticError(resourceName, fmt.Sprintf("Failed to create routing email domain %s error: %s", domainID, err), resp)
+	}
+
+	d.SetId(*domain.Id)
+	log.Printf("Created routing email domain %s", *domain.Id)
+
+	// Other settings must be updated in a PATCH update
+	if d.HasChanges("mail_from_domain", "custom_smtp_server_id") {
+		return updateRoutingEmailDomain(ctx, d, meta)
+	} else {
+		return readRoutingEmailDomain(ctx, d, meta)
+	}
+}
+
+func readRoutingEmailDomain(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*provider.ProviderMeta).ClientConfig
+	proxy := getRoutingEmailDomainProxy(sdkConfig)
+
+	cc := consistency_checker.NewConsistencyCheck(ctx, d, meta, ResourceRoutingEmailDomain(), constants.DefaultConsistencyChecks, resourceName)
+
+	log.Printf("Reading routing email domain %s", d.Id())
+	return util.WithRetriesForRead(ctx, d, func() *retry.RetryError {
+		domain, resp, getErr := proxy.getRoutingEmailDomainById(ctx, d.Id())
+		if getErr != nil {
+			if util.IsStatus404(resp) {
+				return retry.RetryableError(util.BuildWithRetriesApiDiagnosticError(resourceName, fmt.Sprintf("Failed to read routing email domain %s | error: %s", d.Id(), getErr), resp))
+			}
+			return retry.NonRetryableError(util.BuildWithRetriesApiDiagnosticError(resourceName, fmt.Sprintf("Failed to read routing email domain %s | error: %s", d.Id(), getErr), resp))
+		}
+
+		resourcedata.SetNillableValue(d, "subdomain", domain.SubDomain)
+
+		if domain.SubDomain != nil && *domain.SubDomain {
+			// Strip off the regional domain suffix added by the server
+			_ = d.Set("domain_id", strings.SplitN(*domain.Id, ".", 2)[0])
+		} else {
+			_ = d.Set("domain_id", *domain.Id)
+		}
+
+		if domain.CustomSMTPServer != nil && domain.CustomSMTPServer.Id != nil {
```

Comment:
There's a function in resourcedata to handle reading `DomainEntityRef` values you can use here